payment-failure/src/main/scala/com/gu/identity/paymentfailure/abtest/AutoSignInTest.scala

```diff
@@ -17,7 +17,7 @@ class AutoSignInTest(identityClient: IdentityClient) extends VariantGenerator {
           val body = AutoSignInLinkRequestBody(identityId, email)
           identityClient.createAutoSignInToken(body)
             .bimap(
-              err => new RuntimeException("unable to create auto sign-in token", err),
+              err => UserIneligibleForAbTest("unable to create auto sign-in token", Some(err)),
```

Comment:
To send an email, the `BrazeEmailServiceWithAbTest`:
1. generates a variant
2. requests an email be sent including the variant data as meta data

It will only recover from a failure to execute these steps if the `UserIneligibleForAbTest`, in which case it will send an email without any variant meta data; otherwise the error will be propagated and the lambda will be executed again (i.e. retry to send the email with variant data). 

This behaviour means that the variant generator should only return an `UserIneligibleForAbTest` if we explicitly don't want to try and send the email again with the variant meta data, for example, there was a failure from the `IdentityApiClient` which is not likely to change over consecutive invocations of the lambda. 

Conversely, if we think that the error from the `IdentityApiClient` was ephemeral, we want to try and execute the lambda again, so don't return an `UserIneligibleForAbTest`.

Therefore, in the context of this generator I would recommend something like:
```scala
def handleIdentityClientError(err: IdentityClientError): Throwable =
  err match {
    // If the user is invalid for an auto sign-in token, 
    // return an UserIneligibleForAbTest,
    // so that the BrazeEmailServiceWithAbTest will send them a regular email
    case apiError: ApiError if apiError.invalidUser => UserIneligibleForAbTest("user not eligible for auto sign-in token", Some(apiError))
    // Otherwise return a RuntimeException
    // The BrazeEmailServiceWithAbTest will propagate this error,
    // meaning the lambda will be invoked again 
    // i.e. another attempt will be made to send the email to the reader
    // with the variant meta data
    case _ => new RuntimeException("unable to create auto sign-in token", err)
  }

override def generateVariant(identityId: String, email: String): Either[Throwable, Variant] =
  VariantGenerator.getSegmentId(identityId, from = 0, to = 0.2)
    .flatMap {
      case id if id < 0.1 => Right(controlVariant)
      case _ =>
        val body = AutoSignInLinkRequestBody(identityId, email)
        identityClient.createAutoSignInToken(body)
          .bimap(
            handleIdentityClientError,
            response => autoSignInTokenVariant(response.token)
          )
    }
```

Reply:
A couple of remarks regarding this:
1. We could be more specific / declarative about the left hand side of the method `VariantGenerator.generateVariant()` e.g. `VariantGeneratorError` where:

    ```scala
    sealed trait VariantGeneratorError
    case class SendEmailWithoutVariantData(...) extends VariantGeneratorError
    case class RetrySendEmailWithVariantData(...) extends VariantGeneratorError
    ```
2. Or, we could encode the difference between sending without variant data or retrying with variant data by changing the return type of `VariantGenerator.generateVariant()`:

    ```scala
    def generateVariant(): Either[Throwable, Option[Variant]]
    ```
    where a `None` means send email without variant data, `Throwable` retry. In general though, I am weary of stacking monads as it leads to monad transformers or nested higher level method calls e.g. `.flatMap(_.flatMap(...))`

Ultimately, it's about making a decision on what we think is clearest. IMO I _think_ it's ok for now (maybe an additional comment on the `BrazeEmailServiceWithAbTest` to say _if the variant generator returns an error of type `UserIneligibleForAbTest` then..._)

Reply:
Just been through the tests again and see what you mean. I've made the changes you mentioned and added a comment in BrazeEmailServiceWithAbTest for now regarding when we are sending emails or not. I started looking into point 1. (having VariantGeneratorError trait) but I think it's quite clear now with the UserIneligibleForAbTest case class and comments what's happening. What do you think?

Reply:
Yeah, I think it's fine as is. Interesting to think about the alternatives though. At least for me 😂🙈